### ISSUE TITLE
Alternatives to named(array)tuple?

### ISSUE THREAD

**Issue by qxcv** [astooke/rlpyt#99]
Thanks for the great library! I've been trying to use it for an OpenGL env that doesn't like being `fork()`ed. That rules out using the default `fork` context in `multiprocessing`, which I've had to replace with `spawn`. Unfortunately there are several places (Gym env wrapper, Gym space wrapper, `buffer_from_example`, etc.) where `rlpyt` dynamically creates new `namedtuple` or `namedarraytuple` types and stores them as globals in some module. `spawn` does not preserve globals, so those types cannot be found in the child processes created by `multiprocessing`, and the whole thing falls down due to `pickle` errors. Although I haven't tried it, I expect something similar would happen if I tried to pickle buffers (e.g. to store demonstrations to disk) & then reload them in a different interpreter later on. This is a major limitation for my use case.

I've gotten around this problem by replacing `namedtuple` and `namedarraytuple` with [duck-type equivalents](https://github.com/qxcv/rlpyt/blob/b69094b20f08328fe89c358f3271b7c83c360de2/rlpyt/utils/collections.py) that _don't_ create a new type for each possible type name & combination of fields. Roughly it looks like the code below (for `namedtuple`; `namedarraytuple` is similar):

```
class NamedTupleSchema:
    # instances of this class act a bit like a type returned by namedtuple()
    def __init__(self, type_name, fields):
        self.type_name = type_name
        self._fields = fields
    def __call__(self, *values):
        # allows instances of `NamedTupleSchema` to act like `namedtuple` constructors
        return NamedTuple(self, values)

class NamedTuple(tuple):
    # instances of this class act like instances of namedtuple types
    def __new__(cls, schema, values):
        self = tuple.__new__(cls, values)
        self.schema = schema
        return self
    @property
    def __getattr__(self, name):
        # remaining methods (_make, _fields, etc.) are similar
        return self[self.schema._fields.index(name)]
```

This is hacky, but was the path of least resistance (e.g. all the buffer pre-allocation code still works fine without me changing anything outside `collections.py`). What would you think about incorporating a less hacky solution into rlpyt? I was thinking something along the lines of replacing all the `namedtuple` and `namedarraytuple` instances with nests of dicts/tuples/etc. The ergonomics of `namedarraytuple` could be preserved with a helper `dict` wrapper that supports array-based numeric indexing. Perhaps you have better ideas, though.

**Comment by astooke** [comment]
Hi, thanks for the kind words!  And yes, I agree, the current mechanism for dynamically defining namedarraytuples is cumbersome and problematic--I like the idea of looking for an easier solution for this.  I'm partial to a solution which avoids extra dictionaries and keeps things as light-weight as namedtuples, but maybe it's worth sacrificing that for usability.

I'll play with your solution a bit and see how it works, will tinker some more, and will write back here soon!  :)

**Comment by qxcv** [comment]
For posterity: to get `spawn` to work, it's also necessary to change the `np_mp_array` function to return something that actually has a reference to the MP shared array object. [This was my solution](https://github.com/qxcv/rlpyt/blob/61a08e821f9893eb9b484d50062d651a6cca8078/rlpyt/utils/buffer.py#L44-L95). The [current implementation](https://github.com/astooke/rlpyt/blob/35af87255465b3644747294f7fd1ff6045dab910/rlpyt/utils/buffer.py#L43-L48) just returns a plain `numpy` array, but that [multiprocessing cannot pickle and unpickle that properly in general](https://research.wmz.ninja/articles/2018/03/on-sharing-large-arrays-when-using-pythons-multiprocessing.html). With `fork` the old method works fine, but with `spawn` it will silently initialise a separate array with different contents for each child :cry:

**Comment by astooke** [comment]
Yikes!  That sort of silent bug is not fun to suffer.  Thanks for letting us know differences with spawn and fixes.  I’m finally getting back to making a sort of mock namedarraytuple class along your suggestion, will post it soon to see if it works in your case!

Out of curiosity, what is it about the forking that breaks OpenGL?  The environment instances are created inside the workers after they fork.  There is one exception though, an example environment is instantiated before forking, but it’s deleted.  I guess if you have to import opengl in some file header, that could cause the problem?

**Comment by qxcv** [comment]
> Out of curiosity, what is it about the forking that breaks OpenGL? … I guess if you have to import opengl in some file header, that could cause the problem?

Yes, my code was importing graphics libs in the parent process before forking (but not constructing environments). That could have been part of the issue. I don't know enough about the Linux graphics stack to explain why that would cause children to crash later on, though. For what it's worth, the crash only occurred when trying to blit to an FBO in the child; earlier OpenGL calls in the child went through fine.

**Comment by ankeshanand** [comment]
Ran into a similar problem, and I found `mp.Lock()` objects are also not picklable, which is problematic when pickling any of the async buffers while sending them to a child process.

**Comment by astooke** [comment]
@qxcv Here is an attempt at building on your solution and with more safeguards in place to make sure the objects behave like namedtuples, for example making them immutable (it's on a new branch "namedtuple_schema"):

https://github.com/astooke/rlpyt/blob/namedtuple_schema/rlpyt/utils/namedtuple_schema.py

I think the pickling should work, too.  Would you mind giving it a quick try and see if it goes through?

**Comment by qxcv** [comment]
Great, thanks for implementing this! I'm a bit busy today & tomorrow, but I should be able to try it out on Thursday afternoon & update this thread then.

**Comment by astooke** [comment]
@ankeshanand Interesting...what is the purpose of pickling the replay buffer?  Want to start a separate issue?

**Comment by astooke** [comment]
@qxcv I just pushed a few more changes to that branch, which makes all the buffer functionality compatible with either namedtuple or NamedTuple (luckily was very easy change!).  Also made a change to `buffer_from_example()` to accept a kwarg which flags whether to make a NamedArrayTuple or a namedarraytuple, regardless of which kind is inside the input example.  Hopefully it's a relatively easy fix from here to switch that flag and have the buffers all be the new kind.  

I haven't touched the gym env wrappers tho.  I'll get back to that in a couple days if you haven't already done it by then.  :)

**Comment by qxcv** [comment]
Works for me. I made a [PR for the gym wrapper](https://github.com/astooke/rlpyt/pull/114). Would you also like a PR for the [`np_mp_array`](https://github.com/qxcv/rlpyt/blob/master/rlpyt/utils/buffer.py#L44-L95) fix?

**Comment by astooke** [comment]
Great thanks!  Took your gym wrapper PR.  And I only just saw the details of your much earlier NamedTupleSchema implementation...very nice!  Didn't mean to duplicate all that work hah.  So I incorporated a lot of that except for the _global_cache, although that was really nice.  Also I just moved everything into rlpyt.utils.collections to avoid some circular imports.

As for the np_mp_array fix, very slick.  Given that it's quite complicated compared to the original, what do you think about making it a separate function, next to the original, maybe named `np_mp_array_spawn` or something like that?  And at the top of the original function, write:
```python
if mp.get_start_method() == "spawn":
    return np_mp_array_spawn(shape, dtype)
...rest of code
```
If you PR that, I'll merge it to this branch and then merge into master.  I renamed the gym files with `_schema` at the end, so when I merge into master I'll keep the originals.

**Comment by qxcv** [comment]
Sounds good. Just filed #117 with those changes (I also bundled in some other small changes that I've been using to make my own code work).

**Comment by astooke** [comment]
All patched into master with merge commit: 99fb5b68d7ccbf0caa3c528734fe600340afd6b3  which includes #117.  I think that's all done here, to summarize:

- shared memory should now support spawn (previously only fork) in `np_mp_array` (via `np_mp_array_spawn`).

-  pickling/unpickling of data buffers under spawn now supported using e.g. new NamedArrayTuple replacement for namedarraytuple; added new gym env and space wrappers (appended with `_schema`)

- dynamic creation of namedarraytuple types also may be easier using the schema--no need to get into module globals

Let us know if we missed anything!

@ankeshanand if you want to work on the `mp.Lock()` issue some more, let's move that to a new issue?
